simple_parsing/replace.py

```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+import functools
+import logging
+from dataclasses import MISSING, fields, is_dataclass
+
+from simple_parsing.helpers.serialization.serializable import from_dict, to_dict
+from simple_parsing.utils import unflatten_split, DataclassT, dict_union
+
+
+logger = logging.getLogger(__name__)
+
+
+def invoke_post_init_inplace(dataclass: DataclassT, recurse: bool = True) -> None:
+    for f in fields(dataclass):
+        if f.default_factory is not MISSING:
+            if is_dataclass(f.default_factory) or (
+                isinstance(f.default_factory, functools.partial)
+                and is_dataclass(f.default_factory.func)
+            ):
+                if recurse and is_dataclass(getattr(dataclass, f.name)):
+                    invoke_post_init_inplace(getattr(dataclass, f.name), recurse=recurse)
+
+    if getattr(dataclass, "__post_init__", None):
+        dataclass.__post_init__()
+
+
+def replace(
+    dataclass: DataclassT, new_values_dict: dict | None = None, **new_values: dict
+) -> DataclassT:
+    """Replace some values in a dataclass. Also works with nested fields of the dataclass.
+    This is essentially an extension of `dataclasses.replace` that also allows changing the nested
+    fields of dataclasses.
+    ## Examples
+    ```python
+    from __future__ import annotations
+    from dataclasses import dataclass, field
+    @dataclass
+    class A:
+        a: int = 0
+    @dataclass
+    class B:
+       b: str = "b"
+    @dataclass
+    class Config:
+       a_or_b: A | B = field(default_factory=A)
+    config = Config(a_or_b=A(a=1))
+    assert replace(config, {"a_or_b": {"a": 2}}) == Config(a_or_b=A(a=2))
+    assert replace(config, {"a_or_b.a": 2}) == Config(a_or_b=A(a=2))
+    assert replace(config, {"a_or_b": B(b="bob")}) == Config(a_or_b=B(b='bob'))
+    ```
+    """
+    if not is_dataclass(dataclass):
+        raise TypeError("replace() should be called on dataclass instances")
+    
+    if new_values:
+        # Make sure entries of nwe_values are unflattened
+        for k in new_values.keys():
+            v = new_values[k]
+            if isinstance(v, dict):
+                new_values[k] = unflatten_split(v)
+
+    if new_values_dict:
+        new_values = dict(**new_values_dict, **new_values)
+
+    dataclass_dict = to_dict(dataclass, recurse=True)
+    unflatten_new_values = unflatten_split(new_values)
+    new_dataclass_dict = dict_union(dataclass_dict, unflatten_new_values, recurse=True)
+    # NOTE: The drop_extra_fields for from_dict set to true will hide unwanted error for users
+    # Is there any case needs drop_extrac_fields to be True?
+    replaced_dataclass = from_dict(
+        type(dataclass), new_dataclass_dict, drop_extra_fields=False
+    )
+    invoke_post_init_inplace(replaced_dataclass)
+    return replaced_dataclass
```

Comment:
Here is what I believe would be a simpler implementation for this feature, where we don't depend on any of the serialization / deserialization logic of `from_dict` and `to_dict`.
By the way, all your tests work with this. Some of them also XPASS (the last two in `test_replace_post_init`).

Please do let me know what you think.
```suggestion
from __future__ import annotations
import dataclasses

from typing import Any
from simple_parsing.utils import (
    unflatten_split,
    is_dataclass_instance,
    DataclassT,
)


def replace(obj: DataclassT, changes_dict: dict[str, Any] | None = None, **changes) -> DataclassT:
    """Replace some values in a dataclass. Also works with nested fields of the dataclass.
    This is essentially an extension of `dataclasses.replace` that also allows changing the nested
    fields of dataclasses.

    ## Examples
    >>> from dataclasses import dataclass, field
    >>> from typing import Union
    >>> @dataclass
    ... class A:
    ...    a: int = 0
    >>> @dataclass
    ... class B:
    ...     b: str = "b"
    >>> @dataclass
    ... class Config:
    ...     a_or_b: Union[A, B] = field(default_factory=A)

    >>> config = Config(a_or_b=A(a=1))

    Replace accepts either a dictionary of changes or keyword arguments:
    >>> replace(config, {"a_or_b": B(b="bob")})
    Config(a_or_b=B(b='bob'))
    >>> replace(config, a_or_b=B(b='bob'))
    Config(a_or_b=B(b='bob'))

    Changes can also be passed in a 'flat' format, which makes it easy to replace nested fields:
    >>> replace(config, {"a_or_b.a": 2})
    Config(a_or_b=A(a=2))
    """
    if changes_dict and changes:
        raise ValueError("Cannot pass both `changes_dict` and `changes`")
    changes = changes_dict or changes
    # changes can be given in a 'flat' format in `changes_dict`, e.g. {"a.b.c": 123}.
    # Unflatten them back to a nested dict (e.g. {"a": {"b": {"c": 123}}})
    changes = unflatten_split(changes)

    replace_kwargs = {}
    for field in dataclasses.fields(obj):
        if field.name not in changes:
            continue
        if not field.init:
            raise ValueError(f"Cannot replace value of non-init field {field.name}.")

        field_value = getattr(obj, field.name)

        if is_dataclass_instance(field_value) and isinstance(changes[field.name], dict):
            field_changes = changes[field.name]
            new_value = replace(field_value, **field_changes)
        else:
            new_value = changes[field.name]
        replace_kwargs[field.name] = new_value
    return dataclasses.replace(obj, **replace_kwargs)
```

Reply:
Uh, I think we should keep the `replace` function consistant for not checking the type.

Reply:
Indeed, checking the annotation is really difficult (especially with postponed annotations).